### ISSUE TITLE
javascript error: Electron APIs can not be invoked outside of WDIO

### ISSUE THREAD

**Issue by SanjeevForYou** [webdriverio-community/wdio-electron-service#508]
Generated electron project as suggested [here](https://www.electronforge.io/templates/vite-+-typescript) 
`
npm init electron-app@latest my-new-app -- --template=vite-typescript
`

Configured Webdriver for electron as suggested [here](https://webdriver.io/docs/desktop-testing/electron),
`
npm create wdio@latest ./
`

Also, followed the step for accessing electron api step defined [here](https://webdriver.io/docs/desktop-testing/electron/api)

Was trying to run a simple test.e2e.ts with `npm run wdio` command,

```
import { expect } from '@wdio/globals'
import { browser } from 'wdio-electron-service'

describe('Electron Testing', () => {
    it('should open the default window height and width', async () => {
        const bounds = await browser.electron.execute((electron) => {
          const focusedWindow =  electron.BrowserWindow.getFocusedWindow()
          return focusedWindow.getBounds()
        }) 

        await expect(bounds).toEqual(800)
        await expect(bounds).toEqual(600)
    })
})

```

I am constantly getting below error,

```

[0-0] 2024-04-04T22:45:07.307Z WARN webdriver: Request failed with status 500 due to javascript error: Electron APIs can not be invoked outside of WDIO
[0-0]   (Session info: chrome=122.0.6261.156)
[0-0] 2024-04-04T22:45:07.307Z INFO webdriver: Retrying 3/3
[0-0] 2024-04-04T22:45:07.307Z INFO webdriver: [POST] http://0.0.0.0:59944/session/c77850cde9a524af635989307f0a3299/execute/sync
[0-0] 2024-04-04T22:45:07.308Z INFO webdriver: DATA {
[0-0]   script: 'executeWithinElectron(...) [635 bytes]',
[0-0]   args: [
[0-0]     '(electron) => {\n' +
[0-0]       '            const focusedWindow = electron.BrowserWindow.getFocusedWindow();\n' +
[0-0]       '            return focusedWindow.getBounds();\n' +
[0-0]       '        }'
[0-0]   ]
[0-0] }
[0-0] 2024-04-04T22:45:07.639Z ERROR webdriver: Request failed with status 500 due to javascript error: javascript error: Electron APIs can not be invoked outside of WDIO
[0-0]   (Session info: chrome=122.0.6261.156)
[0-0] javascript error in "Electron Testing.should open the default window height and width"
javascript error: javascript error: Electron APIs can not be invoked outside of WDIO
  (Session info: chrome=122.0.6261.156)
    at async ElectronWorkerService.execute (file:///Users/sb/SampleProjects/electron-wdio-vite-typescript/elvitw/node_modules/wdio-electron-service/src/commands/execute.ts:21:23)
    at async Context.<anonymous> (/Users/sb/SampleProjects/electron-wdio-vite-typescript/elvitw/test/specs/test.e2e.ts:6:24)
[0-0] 2024-04-04T22:45:07.657Z INFO webdriver: COMMAND deleteSession()
[0-0] 2024-04-04T22:45:07.658Z INFO webdriver: [DELETE] http://0.0.0.0:59944/session/c77850cde9a524af635989307f0a3299
[0-0] 2024-04-04T22:45:07.709Z INFO webdriver: RESULT null
[0-0] 2024-04-04T22:45:07.709Z INFO webdriver: Kill driver process with PID 22916
[0-0] FAILED in chrome - file:///test/specs/test.e2e.ts
2024-04-04T22:45:07.841Z INFO @wdio/cli:launcher: Run onWorkerEnd hook
2024-04-04T22:45:07.841Z INFO @wdio/cli:launcher: Run onComplete hook
```

Please suggest what is wrong. 

Environment:
Node version: v20.11.1
Npm version: 10.2.4
OS: Mac OS ( 14.4 ) 

Package.json:

```

{
  "name": "elvitw",
  "productName": "elvitw",
  "version": "1.0.0",
  "description": "My Electron application description",
  "main": ".vite/build/main.js",
  "scripts": {
    "start": "electron-forge start",
    "package": "NODE_ENV=test electron-forge package",
    "make": "electron-forge make",
    "publish": "electron-forge publish",
    "lint": "eslint --ext .ts,.tsx .",
    "wdio": "wdio run ./wdio.conf.ts"
  },
  "devDependencies": {
    "@electron-forge/cli": "^7.3.1",
    "@electron-forge/maker-deb": "^7.3.1",
    "@electron-forge/maker-rpm": "^7.3.1",
    "@electron-forge/maker-squirrel": "^7.3.1",
    "@electron-forge/maker-zip": "^7.3.1",
    "@electron-forge/plugin-auto-unpack-natives": "^7.3.1",
    "@electron-forge/plugin-fuses": "^7.3.1",
    "@electron-forge/plugin-vite": "^7.3.1",
    "@electron/fuses": "^1.8.0",
    "@typescript-eslint/eslint-plugin": "^5.62.0",
    "@typescript-eslint/parser": "^5.62.0",
    "@wdio/cli": "^8.35.1",
    "@wdio/local-runner": "^8.35.1",
    "@wdio/mocha-framework": "^8.35.0",
    "@wdio/spec-reporter": "^8.32.4",
    "electron": "29.2.0",
    "eslint": "^8.57.0",
    "eslint-plugin-import": "^2.29.1",
    "ts-node": "^10.9.2",
    "typescript": "~4.5.4",
    "vite": "^5.2.8",
    "wdio-electron-service": "^6.4.0"
  },
  "keywords": [],
  "author": {
    "name": "test",
    "email": "test@test.com"
  },
  "license": "MIT",
  "dependencies": {
    "electron-squirrel-startup": "^1.0.0"
  }
}
```

Thank you for your help.

**Comment by goosewobbler** [comment]
Hi @SanjeevForYou, basically the environment variable `WDIO_ELECTRON` is set [in the service index](https://github.com/webdriverio-community/wdio-electron-service/blob/main/src/index.ts#L18) in order to ensure that the service API is not used outside the service.  It appears that running the service using this Vite base is causing the environment variable check to fail, I haven't been able to work out why yet.  Setting the variable manually so far hasn't worked.

Vite has a specific way of handling environment variables so we will need to look at what can be done here:
https://vitejs.dev/guide/env-and-mode.html#env-variables-and-modes

**Comment by goosewobbler** [comment]
After a bit of investigation I found [this article](https://dev.to/whchi/how-to-use-processenv-in-vite-ho9) which covers the issue.

I implemented the following to make it work:

create `.env` - at the root of the project
```
WDIO_ELECTRON=true
```

in `vite.preload.config.ts`, import `loadEnv` from vite and use it to get the environment variables from the `.env` file:

```ts
const env = loadEnv(buildConfig.mode, process.cwd(), "WDIO_");
```

then add the following definition to the config:

```ts
    define: {
      "process.env.WDIO_ELECTRON": env.WDIO_ELECTRON,
    },
```

e.g.
```ts
import type { ConfigEnv, UserConfig } from "vite";
import { defineConfig, mergeConfig, loadEnv } from "vite";
import { getBuildConfig, external, pluginHotRestart } from "./vite.base.config";

// https://vitejs.dev/config
export default defineConfig((buildConfig) => {
  const env = loadEnv(buildConfig.mode, process.cwd(), "WDIO_");
  const forgeEnv = buildConfig as ConfigEnv<"build">;
  const { forgeConfigSelf } = forgeEnv;
  const config: UserConfig = {
    build: {
      rollupOptions: {
        external,
        // Preload scripts may contain Web assets, so use the `build.rollupOptions.input` instead `build.lib.entry`.
        input: forgeConfigSelf.entry!,
        output: {
          format: "cjs",
          // It should not be split chunks.
          inlineDynamicImports: true,
          entryFileNames: "[name].js",
          chunkFileNames: "[name].js",
          assetFileNames: "[name].[ext]",
        },
      },
    },
    plugins: [pluginHotRestart("reload")],
    define: {
      "process.env.WDIO_ELECTRON": env.WDIO_ELECTRON,
    },
  };

  return mergeConfig(getBuildConfig(forgeEnv), config);
});
```

I'll keep this issue open and update the docs to cover using the service with Vite.

**Comment by goosewobbler** [comment]
@christian-bromann what do you think about this issue?  Vite is very popular and the above steps add a fair bit of weight to the setup process.

Is there another way to limit Electron API usage to the WDIO environment only, without using environment variables?  The other alternative is to remove the guard entirely.  Also, the issue may be redundant after #308.

**Comment by christian-bromann** [comment]
> as a security measure to ensure that the service API is not used outside the service.

Isn't the general recommendation to only include this script when compiling the Electron application for testing? How does this additional check makes it more safer? Another way I could imagine this to work on top of my head is very using application parameter. WDIO can attach a certain parameter to the application call and only allow invoke the context bridge if the parameter is set.

**Comment by goosewobbler** [comment]
> Isn't the general recommendation to only include this script when compiling the Electron application for testing? 

Yes, and I think the guard for the preload script here isn't secure as it simply adds another step for anyone wishing to access the APIs.  The same applies for the application parameter approach - the parameter would be freely accessible on github, so I think we remove the guard here.

**Comment by goosewobbler** [comment]
Released in `6.4.1`.

**Comment by SanjeevForYou** [comment]
Really appreciate it. Thank you so much for the very quick turnaround.
